lib/logstash/outputs/elasticsearch.rb

```diff
@@ -267,18 +271,31 @@ def register
     # the license_checking behaviour in the Pool class is externalized in the LogStash::ElasticSearchOutputLicenseChecker
     # class defined in license_check.rb. This license checking is specific to the elasticsearch output here and passed
     # to build_client down to the Pool class.
-    build_client(LicenseChecker.new(@logger))
+    @client = build_client(LicenseChecker.new(@logger))
 
-    @template_installer = setup_after_successful_connection do
-      discover_cluster_uuid
-      install_template
-      setup_ilm if ilm_in_use?
+    @event_mapper = -> (e) { event_action_tuple(e) }
+    @event_target = -> (e) { e.sprintf(@index) }
+
+    @after_successful_connection_thread = after_successful_connection do
+      finish_register
+      @after_successful_connection.make_true
     end
+
     @bulk_request_metrics = metric.namespace(:bulk_requests)
     @document_level_metrics = metric.namespace(:documents)
     @logger.info("New Elasticsearch output", :class => self.class.name, :hosts => @hosts.map(&:sanitized).map(&:to_s))
   end
 
+  # @override post-register when ES connection established
+  def finish_register
+    synchronize do # due @i-vars being set on the plugin
```

Comment:
so the idea here is to impose a barrier on the field writes - not to prevent concurrent access.

> due @i-vars being set on the plugin

essentially `after_successful_connection` executes on a separate thread and the plugin might be waiting on that thread to complete (from `multi_receive`) by checking a volatile variable. we've been previously setting instance state on a plugin such as `@index = ...` from the after connection thread. in theory there's no visibility guarantee - a mutex should resolve these concerns without resorting several i-vars to atomic references (as there's now more instance state being updated from `finish_register`).

otherwise yeah I understand how this is confusing and probably still not 100% safe.

Reply:
👍🏼 so many layers to think through here, good analysis.